File: 2026/Hard/30.py
from typing import List

class Solution:
  def findSubstring(self, s: str, words: List[str]) -> List[int]:
    words_count_map = {}, word_len = len(words[0]), window_size = word_len * len(words), result = []

    for word in words:
      words_count_map[word] = words_count_map.get(word, 0) + 1

    for start in range(word_len):
      current_map = {}, un_match_words = 0
      
      i = start
      while i < len(s):
        # 0. 提取单词
        word = s[i:i + word_len]

        # 1. 统计当前窗口内这个单词出现次数
        current_map[word] = current_map.get(word, 0) + 1

        if word not in words_count_map or current_map[word] > words_count_map[word]:
          un_match_words += 1

        # 2. 窗口还没达到完整长度，继续扩张
        if i + word_len - window_size < 0:
          i += word_len
          continue

        # 3. 当前窗口合法
        if un_match_words == 0:
          result.append(i + word_len - window_size)

        # 4. 缩窗口
        left_word_index = i + word_len - window_size
        left_word = s[left_word_index:left_word_index + word_len]

        if left_word not in words_count_map or current_map[left_word] > words_count_map[left_word]:
          un_match_words -= 1

        current_map[left_word] -= 1

        i += word_len

    return result

File: 2026/Hard/30.cs
using System;
using System.Collections.Generic;

public class Solution {
  public IList<int> FindSubstring(string s, string[] words) {
    Dictionary<string, int> wordsCountMap = new Dictionary<string, int>();
    int wordLen = words[0].Length, windowSize = wordLen * words.Length;
    List<int> result = new List<int>();

    // 1. 用 Dictionary 记录 words 里每个 word 出现次数
    foreach (string word in words) {
      if (!wordsCountMap.ContainsKey(word)) {
        wordsCountMap[word] = 0;
      }

      wordsCountMap[word]++;
    }

    // 2. 按 wordLen 分组滑窗
    for (int start = 0; start < wordLen; start++) {
      Dictionary<string, int> currentMap = new Dictionary<string, int>();
      int unMatchWords = 0;

      for (int i = start; i < s.Length; i += wordLen) {
        // 0. 提取单词
        // Substring(start, length)
        string word = s.Substring(i, Math.Min(wordLen, s.Length - i));

        // 1. 统计当前窗口内这个 word 出现次数
        if (!currentMap.ContainsKey(word)) {
          currentMap[word] = 0;
        }

        currentMap[word]++;

        if (!wordsCountMap.ContainsKey(word) || currentMap[word] > wordsCountMap[word]) {
          unMatchWords++;
        }

        // 2. 窗口长度还不够，继续扩大
        if (i + wordLen - windowSize < 0) {
          continue;
        }

        // 3. 如果当前窗口没有 mismatch，记录答案
        if (unMatchWords == 0) {
          result.Add(i + wordLen - windowSize);
        }

        // 4. 缩窗口，移除最左边的 word
        int leftWordIndex = i + wordLen - windowSize;
        string leftWord = s.Substring(leftWordIndex, wordLen);

        if (!wordsCountMap.ContainsKey(leftWord) || currentMap[leftWord] > wordsCountMap[leftWord]) {
          unMatchWords--;
        }

        currentMap[leftWord]--;
      }
    }

    return result;
  }
}